scripts/benchmarking.py

```diff
@@ -0,0 +1,210 @@
+
+
+def get_docs(n, n_dim, tensor_shape, n_query):
+    return [
+        Document(
+            embedding=np.random.rand(n_dim),
+            # tensor=np.random.rand(*tensor_shape),
+            tags={'i': int(i / n_query)},
+        )
+        for i in range(n)
+    ]
+
+
+def fmt(value, unit):
+    return '{:.3f} {}'.format(value, unit)
+
+
+def recall(predicted, relevant, eval_at):
+    if eval_at == 0:
+        return 0.0
+    predicted_at_k = predicted[:eval_at]
+    n_predicted_and_relevant = len(
+        set(predicted_at_k[:, 'id']).intersection(set(relevant[:, 'id']))
+    )
+    return n_predicted_and_relevant / len(relevant)
+
+
+if args.default_hnsw:
+    storage_backends = [
+        ('memory', None),
+        ('qdrant', {'n_dim': D, 'scroll_batch_size': 8}),
+        ('sqlite', None),
+        (
+            'annlite',
+            {'n_dim': D},
+        ),
+        ('weaviate', {'n_dim': D}),
+        ('elasticsearch', {'n_dim': D}),
+    ]
+else:
+    storage_backends = [
+        ('memory', None),
+        ('qdrant', {'n_dim': D, 'scroll_batch_size': 8, 'ef_construct': 100, 'm': 16}),
+        ('sqlite', None),
+        (
+            'annlite',
+            {
+                'n_dim': D,
+                'ef_construction': 100,
+                'ef_search': 100,
+                'max_connection': 16,
+            },
+        ),
+        (
+            'weaviate',
+            {'n_dim': D, 'ef': 100, 'ef_construction': 100, 'max_onnections': 16},
+        ),
+        ('elasticsearch', {'n_dim': D, 'ef_construction': 100, 'm': 16}),
+    ]
+
+table = Table(
+    title=f'DocArray Benchmarking n_index={n_index} n_query={n_query} D={D} K={K}'
+)
+table.add_column('Storage Backend')
+table.add_column('Indexing time (C)')
+table.add_column('Query (R)')
+table.add_column('Update (U)')
+table.add_column('Delete (D)')
+table.add_column('Find by vector')
+table.add_column(f'Recall at k={K} for vector search')
+table.add_column('Find by condition')
+
+console = Console()
+
+console.print(f'generating {n_index} docs...')
+docs = get_docs(n_index, D, TENSOR_SHAPE, n_query)
+
+vector_queries = [np.random.rand(n_query, D) for _ in range(n_vector_queries)]
+ground_truth = []
+
+for backend, config in storage_backends:
+    try:
+        console.print('Backend:', backend.title())
+        # for n_i in n_index:
+        if not config:
+            da = DocumentArray(storage=backend)
+        else:
+            da = DocumentArray(storage=backend, config=config)
+
+        console.print(f'indexing {n_index} docs ...')
+        create_time, _ = create(da, docs)
+
+        # for n_q in n_query:
+        console.print(f'reading {n_query} docs ...')
+        read_time, _ = read(
+            da,
+            random.sample([d.id for d in docs], n_query),
+        )
+
+        console.print(f'updating {n_query} docs ...')
+        update_time, _ = update(da, random.sample(docs, n_query))
+
+        docs_to_delete = random.sample(docs, n_query)
```

Comment:
`docs_to_delete` and `docs_to_update` should be same across different backends. The random delete and update here make the ground-truth from memory backend is not correct for other backends. That would be the reason why `weaviate` and `annlite` get poor recall result.

Reply:
Thanks for pointing this out @numb3r3 
I'll update the table with new results once ready

Reply:
table updated with fixed script